File: dsa-problems/Concurrency/solutions/python/CON-006-deadlock-detection-resource-graph.py

def has_deadlock(n: int, edges: list[tuple[int, int]]) -> bool:
    # Nodes 1..n
    adj = [[] for _ in range(n + 1)]
    in_degree = [0] * (n + 1)
    
    for u, v in edges:
        adj[u].append(v)
        in_degree[v] += 1
        
    queue = []
    for i in range(1, n + 1):
        if in_degree[i] == 0:
            queue.append(i)
            
    processed_count = 0
    while queue:
        u = queue.pop(0)
        processed_count += 1
        
        for v in adj[u]:
            in_degree[v] -= 1
            if in_degree[v] == 0:
                queue.append(v)
                
    return processed_count != n

def main():
    import sys
    # Use generator pattern for robust reading
    def input_gen():
        for line in sys.stdin:
            for token in line.split():
                yield token
    it = input_gen()
    
    try:
        n = int(next(it))
        m = int(next(it))
        edges = []
        for _ in range(m):
            u = int(next(it))
            v = int(next(it))
            edges.append((u, v))
        
        if has_deadlock(n, edges):
            print("DEADLOCK")
        else:
            print("NO DEADLOCK")
    except StopIteration:
        pass

if __name__ == "__main__":
    main()

File: dsa-problems/Concurrency/solutions/javascript/CON-006-deadlock-detection-resource-graph.js
const readline = require("readline");

class Solution {
  hasDeadlock(n, edges) {
    const adj = Array.from({ length: n }, () => []);
    const inDegree = new Int32Array(n).fill(0);
    
    for (const [u, v] of edges) {
      adj[u].push(v);
      inDegree[v]++;
    }
    
    const queue = [];
    for (let i = 0; i < n; i++) {
      if (inDegree[i] === 0) {
        queue.push(i);
      }
    }
    
    let head = 0;
    let processedCount = 0;
    
    while (head < queue.length) {
      const u = queue[head++];
      processedCount++;
      
      for (const v of adj[u]) {
        inDegree[v]--;
        if (inDegree[v] === 0) {
          queue.push(v);
        }
      }
    }
    
    return processedCount < n;
  }
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => data.push(...line.trim().split(/\s+/)));
rl.on("close", () => {
  if (data.length === 0) return;
  
  let ptr = 0;
  const n = parseInt(data[ptr++], 10);
  const m = parseInt(data[ptr++], 10);
  
  const edges = [];
  for (let i = 0; i < m; i++) {
    const u = parseInt(data[ptr++], 10);
    const v = parseInt(data[ptr++], 10);
    edges.push([u, v]);
  }
  
  const solution = new Solution();
  console.log(solution.hasDeadlock(n, edges));
});
